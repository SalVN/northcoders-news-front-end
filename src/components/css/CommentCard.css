.comment-card-votes {
    text-align: center;
    font-weight: strong;
}

.comment-card-name {
    color: #c1002e !important;
}