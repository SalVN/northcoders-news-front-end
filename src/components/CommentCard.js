import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';
import moment from 'moment';

import { USERNAME } from '../../config';

import './css/CommentCard.css';

const CommentCard = function (props) {
  return (
    <div className='card box-style'>
      <article className='media'>
        <div className='media-left'>
          <button onClick={props.voteHandler.bind(this, 'up', props.comment._id)} className='button is-black is-inverted is-small up'><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
          <p className='comment-card-votes'><strong>{props.comment.votes}</strong></p>
          <button onClick={props.voteHandler.bind(this, 'down', props.comment._id)} className='button is-black is-inverted is-small down'><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
        </div>
        <div className='media-content'>
          <div className='content'>
              <div>
                <img className='comment-card-avatar' src={props.userData.avatar_url} alt='user avatar' />
                <Link to={`/users/${props.userData.username}`}><span className='comment-card-name'><strong>{props.comment.created_by}</strong></span></Link>
                <p className='comment-card-date'>{moment(props.comment.created_at).utc().format('Do MMMM YYYY, hh:mm')}</p>
              </div>
            <div className='comment-card-body'>{props.comment.body}</div>
            {props.comment.created_by === USERNAME &&
              <button onClick={props.deleteHandler.bind(this, props.comment._id)} className='button is-small delete-button'>Delete Comment</button>
            }
          </div>
        </div>
      </article>
    </div>
  );
};

CommentCard.propTypes = {
  comment: PropTypes.object.isRequired,
  deleteHandler: PropTypes.func.isRequired,
  voteHandler: PropTypes.func.isRequired,
  userData: PropTypes.object.isRequired
};


export default CommentCard;
