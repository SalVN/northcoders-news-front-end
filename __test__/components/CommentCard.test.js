import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';

import CommentCard from '../../src/components/CommentCard';

describe('CommentCard', () => {
    const comment = {
        _id: '59b11ae18807841d9bf13234',
        body: 'this is a comment',
        belongs_to: '59b11ae18807841d9bf13232',
        __v: 0,
        created_by: 'northcoder',
        votes: 1,
        created_at: 1504778965845
    };

    const otherUserComment = {
        _id: '59b11ae18807841d9bf13234',
        body: 'this is a comment',
        belongs_to: '59b11ae18807841d9bf13232',
        __v: 0,
        created_by: 'not_northcoder',
        votes: 1,
        created_at: 1504778965845
    };

    const user = {
        _id: '59b1b18b327cce1fb043bdb1',
        username: 'northcoder',
        name: 'Awesome Northcoder',
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    };

    it('is a function', () => {
        expect(typeof CommentCard).toEqual('function');
    });

    it('renders', () => {
        const wrapper = shallow(<CommentCard
            voteHandler={(x) => { return x; }}
            deleteHandler={(x) => { return x; }}
            comment={comment}
            userData={user}
        />);
        expect(wrapper.children().length).toEqual(1);
    });

    it('renders correctly', () => {
        const tree = renderer.create(
            <MemoryRouter>
                <CommentCard
                    voteHandler={x => x}
                    deleteHandler={x => x}
                    comment={comment}
                    userData={user}
                />
            </MemoryRouter>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('contains one link', () => {
        const wrapper = shallow(<CommentCard
            voteHandler={x => x}
            deleteHandler={x => x}
            comment={comment}
            userData={user}
        />);
        expect(wrapper.find('Link').length).toBe(1);
    });

    it('renders a delete button if the comment was added by the user', () => {
        const wrapper = shallow(<CommentCard
            voteHandler={(x) => { return x; }}
            deleteHandler={(x) => { return x; }}
            comment={comment}
            userData={user}
        />);
        expect(wrapper.find('.delete-button').length).toEqual(1);
    });

    it('does not render a delete button if the comment was added by a different user', () => {
        const wrapper = shallow(<CommentCard
            voteHandler={(x) => { return x; }}
            deleteHandler={(x) => { return x; }}
            comment={otherUserComment}
            userData={user}
        />);
        expect(wrapper.find('.delete-button').length).toEqual(0);
    });

    it('calls the voteHandler function when the "up" button is clicked', () => {
        const spy = sinon.stub();
        const wrapper = shallow(<CommentCard
            voteHandler={spy}
            deleteHandler={(x) => { return x; }}
            comment={comment}
            userData={user}
        />);
        expect(spy.called).toBe(false);
        wrapper.find('.up').simulate('click', { preventDefault() { } });
        expect(spy.called).toBe(true);
    });

    it('calls the voteHandler function when the "down" button is clicked', () => {
        const spy = sinon.stub();
        const wrapper = shallow(<CommentCard
            voteHandler={spy}
            deleteHandler={(x) => { return x; }}
            comment={comment}
            userData={user}
        />);
        expect(spy.called).toBe(false);
        wrapper.find('.down').simulate('click', { preventDefault() { } });
        expect(spy.called).toBe(true);
    });

    it('calls the deleteHandler function when the "delete" button is clicked', () => {
        const spy = sinon.stub();
        const wrapper = shallow(<CommentCard
            voteHandler={x => x}
            deleteHandler={spy}
            comment={comment}
            userData={user}
        />);
        expect(spy.called).toBe(false);
        wrapper.find('.delete-button').simulate('click', { preventDefault() { } });
        expect(spy.called).toBe(true);
    });
});