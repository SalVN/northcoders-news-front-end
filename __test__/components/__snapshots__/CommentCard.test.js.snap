// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentCard renders correctly 1`] = `
<div
  className="card box-style"
>
  <article
    className="media"
  >
    <div
      className="media-left"
    >
      <button
        className="button is-black is-inverted is-small up"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-arrow-up"
        />
      </button>
      <p
        className="comment-card-votes"
      >
        <strong>
          1
        </strong>
      </p>
      <button
        className="button is-black is-inverted is-small down"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-arrow-down"
        />
      </button>
    </div>
    <div
      className="media-content"
    >
      <div
        className="content"
      >
        <div>
          <img
            alt="user avatar"
            className="comment-card-avatar"
            src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
          />
          <a
            href="/users/northcoder"
            onClick={[Function]}
          >
            <span
              className="comment-card-name"
            >
              <strong>
                northcoder
              </strong>
            </span>
          </a>
          <p
            className="comment-card-date"
          >
            7th September 2017, 10:09
          </p>
        </div>
        <div
          className="comment-card-body"
        >
          this is a comment
        </div>
        <button
          className="button is-small delete-button"
          onClick={[Function]}
        >
          Delete Comment
        </button>
      </div>
    </div>
  </article>
</div>
`;
